R/estimateDiversity.R

```diff
@@ -178,13 +188,123 @@ setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
                                          .get_diversity_values,
                                          x = x,
                                          mat = assay(x, abund_values),
+                                         tree = tree,
                                          BPPARAM = BPPARAM,
                                          ...)
         .add_values_to_colData(x, dvrsts, name)
     }
 )
 
+#' @rdname estimateDiversity
+#' @export
+setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
+    function(x, abund_values = "counts",
+             index = c("shannon","gini_simpson","inverse_simpson",
+                       "coverage", "fisher", "faith"),
+             name = index, ..., BPPARAM = SerialParam()){
+        
+        # Gets the tree 
+        tree <- rowTree(x)
+        
+        # input check
+        # If object does not have a tree
+        if( ("faith" %in% index) && (is.null(tree) || is.null(tree$edge.length)) ){
+            stop("Object does not have a tree or the tree does not have any branches.
+             'faith' is not possible to calculate.",
+                 call. = FALSE)
+        }
+        
+        index<- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                 "same length than 'index'.",
+                 call. = FALSE)
+        }
+        
+        # If 'faith' is one of the indices
+        if( "faith" %in% index ){
+            # Get the name of "faith" index
+            faith_name <- name[index %in% "faith"]
+            # And delete it from name
+            name <- name[!index %in% "faith"]
+
+            # Delete "faith" from indices
+            index <- index[!index %in% "faith"]
+            
+            # Faith will be calculated
+            calc_faith <- TRUE
+        } else{
+            # Faith will not be calculated
+            calc_faith <- FALSE
+        }
+
+        # If index list contained other than 'faith' index, the length of the list is over 0
+        if( length(index)>0){
+            # Calculates all indices but not 'faith'
+            x <- callNextMethod()
+        }
+
+        # If 'faith' was one of the indices, 'calc_faith' is TRUE
+        if( calc_faith ){
+            # Calculates faith
+            x <- estimateFaith(x, tree = tree, name = faith_name, ...)
+        }
 
+        return(x)
+
+    }
+)
+
+#' @rdname estimateDiversity
+#' 
+#' @param tree A phylogenetic tree that is used to calculate 'faith' index.
+#' By default, it is \code{rowTree(x)}.
+#' 
+#' @export
+setGeneric("estimateFaith",signature = c("x", "tree"),
+           function(x, tree = "missing", abund_values = "counts",
```

Comment:
The missingness cannot be defined in the generic.

Currently this defines a character vector with the value `missing` as default.

Reply:
Should it be defined somewhere? That was just an attempt to fix a warning that said `tree is not defined`

Reply:
I removed the definition from the current version

Reply:
That warning was the result of line 191 contain the tree variable, which is not defined in that function (and also is not used)

Reply:
Ok, clear